### PR TITLE
feat(speculative): add DFlash draft-model training recipe

### PR DESCRIPTION
**Pull request by khazic**
## What

Adds **DFlash** speculative-decoding draft-model training. DFlash drafts a whole block of tokens **in parallel** via MASK-token denoising conditioned on the frozen target's hidden states, in contrast to EAGLE's autoregressive single-step drafting. Ported from SpecForge — see [sgl-project/SpecForge#465](https://github.com/sgl-project/SpecForge/issues/465) (Qwen3-8B DFlash reference).

It is intentionally a **separate vertical alongside EAGLE**, not folded into the EAGLE recipes — the draft model, block attention mask, loss, and inference differ fundamentally. A shared `SpeculativeDraftRecipe` base for the common scaffolding (loop / checkpointing / target capture) is a sensible follow-up but out of scope here.

## Layout

Under `nemo_automodel/components/speculative/dflash/`:
- `draft_qwen3.py` — non-causal Qwen3 draft whose keys/values are `[target-hidden context | noise block]`, with multi-layer target-feature fusion via `fc`; includes block-parallel `spec_generate` inference.
- `core.py` — `DFlashTrainerModule`: per-sequence anchor sampling, MASK noise construction, the DFlash block attention mask (`flex_attention` + SDPA backends), and the block-wise cross-entropy loss with positional decay.
- `target.py` — `HFDFlashTargetModel`: online capture of the selected target decoder layers' hidden states as draft context.
- `registry.py` — architecture → draft dispatch (Qwen3 dense / MoE).

`recipes/llm/train_dflash.py` (`TrainDFlashRecipe`) mirrors the EAGLE recipes' scaffolding (online target capture, grad-accum + trailing flush, the same checkpointer plumbing) plus step/epoch/final checkpoint cadence. A single `attention_backend` knob drives both the mask format and the draft attention fn. Short conversations (fewer than `block_size + 1` supervised tokens) are skipped per micro-batch rather than aborting the run.

Example config at `examples/speculative/dflash/qwen3_dflash.yaml` (SpecForge Qwen3-8B demo hyperparameters: block_size 16, num_anchors 512, loss_decay_gamma 7.0).

## Validation

**Unit tests** (new, CPU): block-mask correctness, trainer loss/grad, draft forward, registry, recipe checkpoint cadence. Full speculative suite: 224 passed / 3 skipped (30 new), `ruff` + format clean.

**Live training smoke** — Qwen3-4B target, real chat data, single GPU, SDPA backend, warmup disabled to see learning quickly. The draft learns end-to-end (loss from the random-init value down, block-prediction accuracy rising from 0):

| step | loss | acc |
|---|---|---|
| 1 | 12.74 | 0.00 |
| 20 | 8.06 | 0.03 |
| 100 | 7.81 | 0.03 |
| 500 | 7.07 | 0.05 |
| 1000 | 6.31 | 0.08 |
| 1500 | 6.73 | 0.06 |
| 1910 | 5.02 | 0.13 |

Run settings:
```
python -m nemo_automodel.recipes.llm.train_dflash --config examples/speculative/dflash/qwen3_dflash.yaml \
  --recipe_args.target_model_name_or_path=Qwen/Qwen3-4B \
  --recipe_args.train_data_path=<chat.jsonl> \
  --recipe_args.seq_length=512 --recipe_args.micro_batch_size=1 --recipe_args.num_anchors=32 \
  --recipe_args.num_workers=0 --recipe_args.attention_backend=sdpa --optimizer.warmup_ratio=0.0
```

> Note: the SDPA backend is validated above; `flex_attention` (the GPU main path) is the default and still needs a hardware validation pass. `num_workers>0` currently hits a tokenizer-pickle issue in the shared chat dataloader (use `num_workers=0`); that's a pre-existing dataloader concern, tracked separately.

### CODE REVIEW

**Comment by khazic** [comment]
Note on overlap with #2315 (already merged): that PR added DFlash under the diffusion-LM (`dllm`) line — `DiffusionLMSFTRecipe` plus `nemo_automodel/components/attention/dflash_mask.py` — targeting **SFT of an existing DFlash model** (e.g. `z-lab/Qwen3-4B-DFlash-b16`).

This PR covers a **different workflow**: training a DFlash **draft from scratch** against a frozen target model online (the SpecForge approach, sgl-project/SpecForge#465), mirroring the EAGLE recipes under `nemo_automodel/components/speculative/`.

There is real overlap — most notably the block attention mask here duplicates `components/attention/dflash_mask.py`. My proposal: land this PR first, then I'll do a dedicated follow-up to reuse the shared mask/loss and cut the duplicated speculative-DFlash code. Happy to adjust if you'd prefer the consolidation up front.

File: nemo_automodel/recipes/llm/train_dflash.py [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+                    self.optimizer.step()
+                    self.optimizer.zero_grad(set_to_none=True)
+                    self.lr_scheduler.step()
+                    self.runtime.global_step += 1
+                    completed_steps += 1
+                    pending_micro_batches = 0
+                    self._maybe_save_step_checkpoint(epoch_idx)
+
+                    if self.dist_env.is_main and self.runtime.global_step % self.log_every_steps == 0:
+                        logger.info(
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/recipes/llm/train_dflash.py:584]: Fixed in 3c0d95b: the logged loss/acc now average over the micro-batches since the last log (a running_micro counter reset alongside running_loss/running_acc), so grad_accumulation_steps no longer inflates them and skipped short micro-batches are excluded.

File: nemo_automodel/components/speculative/dflash/core.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        mask_context = is_context & (kv_idx < anchor_pos)
+
+        is_draft = kv_idx >= S
+        kv_block_id = (kv_idx - S) // block_size
+        mask_draft = is_draft & (q_block_id == kv_block_id)
+
+        is_valid_block = block_keep_mask[b, safe_q_block_id]
+        in_bounds = q_block_id < N
+        return (mask_context | mask_draft) & is_valid_block & in_bounds
+
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/components/speculative/dflash/core.py:133]: Good catch, the -1 was not intentional. Dropped it in 3c0d95b. The valid mask already restricts anchors to positions with room for a full block, so every valid position is a legitimate anchor; a single valid anchor now works and the spurious NoValidAnchorsError is gone.

File: nemo_automodel/components/speculative/dflash/target.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/components/speculative/dflash/target.py:1]: Added tests/unit_tests/speculative/test_dflash_target.py in 3c0d95b: covers layer-id validation, ModuleList/ModuleDict layer dispatch, the hook capture + feature concatenation, and HF-flag dropping for custom backbones.

**Review by HuiyingLi — APPROVED** [review]

**Comment by kashif** [comment]
@khazic i can take the task of removing duplicate if you like?

**Comment by khazic** [comment]
@kashif If you've got the bandwidth, that'd be totally fine. Go for it, thanks!

**Comment by kashif** [comment]
@khazic already cc'ed you on the PR
